### PR TITLE
Added startup check to fail the router when paths with audit policy is …

### PR DESCRIPTION
**Pull request by yaojiefeng**
Added a check before we start the router to make sure all the paths are getting scanned for audit logging. 

Builds: http://builds.cask.co/browse/CDAP-RUT609-JOB1-2

### CODE REVIEW

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/ExceptedNumberOfAuditPolicyPaths.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+package co.cask.cdap.gateway.router;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Expected number of paths annotated with {@link co.cask.cdap.common.security.AuditPolicy}
+ */
+public final class ExceptedNumberOfAuditPolicyPaths {
+  public static final int EXPECTED_PATH_NUMBER = 86;
```
Comment (albertshau) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/ExceptedNumberOfAuditPolicyPaths.java]: this seems very brittle

**Review by albertshau — COMMENTED** [review]
I don't really understand why we need to add this runtime check? Isn't the number of endpoints a static number based on our code, meaning the unit test is enough? But I also don't get why we need to unit test it, what are we trying to prevent?

**Comment by yaojiefeng** [comment]
Discussed this with @anew before. For audit logging, we scan the classpath to filter out all the handler classes and find all the endpoints with annotation @AuditPolicy. So we need to make sure we get all the required endpoints in the classpath. And if some classes are missing in the classpath on a cluster, we will not get any error message but simply not get these endpoints logged without this check.

**Comment by albertshau** [comment]
These are our handler classes right? How can they be missing from the classpath? It means they are missing from our packaging?

**Comment by yaojiefeng** [comment]
Normally this will be no problem. But if someone does something bad on the cluster and eventually we miss some endpoints, it is hard for us to find out because no exception, and no error messages will appear in the log.

**Comment by albertshau** [comment]
It just seems really brittle. Anytime we add or remove an endpoint we need to update that number.  And all this is to protect against somebody going into the router's lib directory and deleting classes?

**Comment by yaojiefeng** [comment]
Yeah, it is brittle. Anytime we add/remove annotation @AuditPolicy to/from endpoints, that number is required to update. I will discuss with Andreas tomorrow to see whether it is worth to do this check.

**Comment by anew** [comment]
@albertshau do you have a better way to ensure that auditing happens for all endpoints? The major concern is that someone may add a new handler in a new maven submodule. Because unit tests load all classes in a single class loader, the unit test will see the new handler and execute correctly for the new endpoints. But in a cluster, we may not inject that handler into the router's class path, and it may not perform the audit logging correctly for this new handler. 

The idea was to have this number in the code to ensure that unit tests and production router see the same set of handlers. We did not like it much, because it means this number must be updated manually to make unit tests pass, but we could not come up with a better way. 

Do you have a better idea?

**Comment by albertshau** [comment]
ok got it, didn't get that it was protecting against new submodules

File: cdap-common/src/main/resources/cdap-default.xml [hunk 2]
```diff
@@ -2401,6 +2401,14 @@
     </description>
   </property>
 
+  <property>
+    <name>router.audit.path.check.enabled</name>
```
Comment (albertshau) [inline cdap-common/src/main/resources/cdap-default.xml]: settings in cdap-default.xml need to be in alphabetical order, and you need to update DEFAULT_XML_MD5_HASH in cdap-docs/admin-manual/build.sh with the new md5 of cdap-default.xml

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/ExceptedNumberOfAuditPolicyPaths.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+
+package co.cask.cdap.gateway.router;
+
+import com.google.common.annotations.VisibleForTesting;
```
Comment (albertshau) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/ExceptedNumberOfAuditPolicyPaths.java]: unused

File: cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java [hunk 4]
```diff
@@ -152,11 +155,15 @@ private HttpMethod getHttpMethod(Method method) {
...
-
       if (Classes.isSubTypeOf(info.getName(), HttpHandler.class.getName(), lookup, cache)) {
         results.add(info.load());
       }
     }
     return results;
   }
+
+  @VisibleForTesting
+  public int getNumberOfPaths() {
```
Comment (albertshau) [inline cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterAuditLookUp.java]: this can be package private

**Review by albertshau — COMMENTED** [review]
small comments

**Comment by yaojiefeng** [comment]
Quick doc build passed: http://builds.cask.co/browse/CDAP-DQB283-3

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**